hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java

```diff
@@ -0,0 +1,408 @@
+        this.eventJournalReader = eventJournalReader;
+        this.partitions = partitions;
+        this.predicate = predicate != null ? predicate::test : null;
+        this.projection = createProjection(projectionF);
+        this.startFromLatestSequence = startFromLatestSequence;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws Exception {
+        jobFuture = context.jobFuture();
+
+        List<ICompletableFuture<EventJournalInitialSubscriberState>> futures = new ArrayList<>();
+        for (Integer partitionId : partitions) {
+            futures.add(eventJournalReader.subscribeToEventJournal(partitionId));
+        }
+        for (int i = 0; i < futures.size(); i++) {
+            ICompletableFuture<EventJournalInitialSubscriberState> future = futures.get(i);
+            offsetMap.put(partitions.get(i), getSequence(future.get()));
+        }
+    }
+
+    private long getSequence(EventJournalInitialSubscriberState state) {
+        if (startFromLatestSequence) {
+            return state.getNewestSequence() == -1 ? 0 : state.getNewestSequence();
+        }
+        return state.getOldestSequence();
+    }
+
+    @Override
+    public boolean complete() {
+        long idleCount = 0;
+        while (!jobFuture.isDone()) {
+            boolean fetched = poll();
+            if (fetched) {
+                idleCount = 0;
+            } else {
+                IDLER.idle(++idleCount);
+            }
+        }
+        return true;
+    }
+
+    private boolean poll() {
+        Map<Integer, ICompletableFuture<ReadResultSet<T>>> futureMap = new HashMap<>();
+        for (Map.Entry<Integer, Long> entry : offsetMap.entrySet()) {
+            int partitionId = entry.getKey();
+            long offset = entry.getValue();
+            ICompletableFuture<ReadResultSet<T>> future = eventJournalReader.readFromEventJournal(offset,
+                    MIN_FETCH_SIZE, MAX_FETCH_SIZE, partitionId, predicate, projection);
+            futureMap.put(partitionId, future);
+        }
+        boolean fetched = false;
+        for (Map.Entry<Integer, ICompletableFuture<ReadResultSet<T>>> entry : futureMap.entrySet()) {
+            ICompletableFuture<ReadResultSet<T>> future = entry.getValue();
+            try {
+                ReadResultSet<T> resultSet = future.get();
+                for (T event : resultSet) {
+                    emit(event);
+                    fetched = true;
+                }
+                offsetMap.compute(entry.getKey(), (k, v) -> v + resultSet.readCount());
+            } catch (ExecutionException e) {
+                if (e.getCause() instanceof StaleSequenceException) {
+                    long headSeq = ((StaleSequenceException) e.getCause()).getHeadSeq();
+                    long oldOffset = offsetMap.put(entry.getKey(), headSeq);
+                    getLogger().severe("Stale sequence, requested: " + oldOffset + ", currentHead: " + headSeq);
```

Comment:
You also have a somewhat ok message in `((StaleSequenceException) e.getCause()).getMessage()` if you like